doc/man/man3/ChatController.3
.TH "ChatController" 3 "Wed Dec 7 2016" "Version 1.0.0" "Jobify" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ChatController \- 
.SH SYNOPSIS
.br
.PP
.PP
\fC#include <ChatController\&.h>\fP
.PP
Inherits JsonController\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "\fBChatController\fP (\fBDatabaseManager\fP *pManager, \fBSessionManager\fP *pSessionManager, std::string fireServerToken)"
.br
.ti -1c
.RI "void \fBsetup\fP ()"
.br
.ti -1c
.RI "void \fBpostUserChatRequest\fP (Mongoose::Request &request, Mongoose::JsonResponse &response)"
.br
.ti -1c
.RI "void \fBgetUserChatsRequest\fP (Mongoose::Request &request, Mongoose::JsonResponse &response)"
.br
.ti -1c
.RI "void \fBdeleteConversationRequest\fP (Mongoose::Request &request, Mongoose::JsonResponse &response)"
.br
.ti -1c
.RI "void \fBdeleteMessageRequest\fP (Mongoose::Request &request, Mongoose::JsonResponse &response)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Controller that manages the chat related api endpoints\&. Every method recieves a Moongose::Request with the request data and a Mongoose::JsonResponse to store the response json\&. 
.SH "Member Function Documentation"
.PP 
.SS "void ChatController::deleteConversationRequest (Mongoose::Request & request, Mongoose::JsonResponse & response)"
Endpoint to delete a conversation with another user 
.PP
\fBParameters:\fP
.RS 4
\fIrequest\fP The request data 
.br
\fIresponse\fP The JsonResponse to send 
.RE
.PP

.SS "void ChatController::deleteMessageRequest (Mongoose::Request & request, Mongoose::JsonResponse & response)"
Endpoint to delete a chat message with another user 
.PP
\fBParameters:\fP
.RS 4
\fIrequest\fP The request data 
.br
\fIresponse\fP The JsonResponse to send 
.RE
.PP

.SS "void ChatController::getUserChatsRequest (Mongoose::Request & request, Mongoose::JsonResponse & response)"
Endpoint to get chat messages with another user 
.PP
\fBParameters:\fP
.RS 4
\fIrequest\fP The request data 
.br
\fIresponse\fP The JsonResponse to send 
.RE
.PP

.SS "void ChatController::postUserChatRequest (Mongoose::Request & request, Mongoose::JsonResponse & response)"
Endpoint to send a chat message to another user 
.PP
\fBParameters:\fP
.RS 4
\fIrequest\fP The request data 
.br
\fIresponse\fP The JsonResponse to send 
.RE
.PP


.SH "Author"
.PP 
Generated automatically by Doxygen for Jobify from the source code\&.


src/ChatController.cpp
//
// Created by nicolas on 26/11/16.
//
// TOKEN mocca@gmail.com = bW9jY2FAZ21haWwuY29tOjEyMzQ=

#include <exceptions/TokenInvalidException.h>
#include <log/easylogging++.h>

#include "ChatController.h"
#include "exceptions/KeyDoesntExistException.h"
#include "ApiError.h"

using Json::Value;
using namespace Mongoose;


ChatController::ChatController(DatabaseManager *db, SessionManager *sessionManager, std::string fireServerToken)
        : db(db), sessionManager(sessionManager) {
    notSender = new NotificationSender(fireServerToken);
}

void ChatController::setup() {
    addRouteResponse("GET", "/chats", ChatController, getUserChatsRequest, JsonResponse);
    addRouteResponse("POST", "/chats", ChatController, postUserChatRequest, JsonResponse);
    addRouteResponse("DELETE", "/chats", ChatController, deleteConversationRequest, JsonResponse);
    addRouteResponse("DELETE", "/messages", ChatController, deleteMessageRequest, JsonResponse);
}

void ChatController::getUserChatsRequest(Mongoose::Request &request, Mongoose::JsonResponse &response) {
    std::string token = request.getHeaderKeyValue("Token");
    std::string username;

    try {
        username = sessionManager->get_username(token);
        std::string username2 = request.get("user","");

        LOG(INFO) << "USER CHAT GET REQUEST:\n"
                  << "\t\tHeader Token: " << token << "\n"
                  << "\t\tUser: " << username
                  <<"\t\tConversacion con: " << username2 << std::endl;

        Json::Value user = db->get_user(username);
        Json::Value conversation;

        if (user["username"] != username) {
            ApiError::setError(response,500,"Internal server error");
        } else {
            bool ok = false;
            if (username2 == "") {
                ok = db->get_convs(username, &conversation);
            } else {
                Json::Value user2 = db->get_user(username2);
                if (user2["username"] != username2) {
                    ApiError::setError(response, 500, "Internal server error");
                } else {
                    ok = db->get_conv(username, username2, &conversation);
                }
            }
            if (!ok) {
                ApiError::setError(response,500,"Internal server error");
            } else {
                response[STATUS] = SUCCES;
                if (username2 == "") {
                    response[DATA] = conversation["chats"];
                } else {
                    response[DATA] = conversation;
                }
            }
        }
    } catch (TokenInvalidException &e) {
        ApiError::setError(response,501,"Token invalido");
    } catch (KeyDoesntExistException &e) {
        ApiError::setError(response,500,"Usuario inexistente");
    }
    LOG(INFO) << "USER CHAT GET RESPONSE:\n"
              << "\t\tUser: " << username << "\n"
              << "\t\tResponse: " << response
              << std::endl;
}

void ChatController::postUserChatRequest(Mongoose::Request &request, Mongoose::JsonResponse &response) {
    std::string token = request.getHeaderKeyValue("Token");
    std::string username;
    std::string data;

    try {
        username = sessionManager->get_username(token);

        data = request.getData();
        Json::Reader reader;
        Json::Value content;
        bool parsingSuccessful = reader.parse(data, content);
        if (!parsingSuccessful) {
            ApiError::setError(response,410,"Wrong JSON");  // TODO agregar a la API documentation
        }
        std::string msgTo = request.get("user","");
        std::string message = content.get("text","").asString();

        LOG(INFO) << "USER CHAT POST REQUEST:\n"
                  << "\t\tHeader Token: " << token << "\n"
                  << "\t\tUser: " << username
                <<"\t\tMensaje a: " << msgTo 
                <<"\t\tMensaje: " << message << "-------" << std::endl;

        Json::Value user = db->get_user(username);
        Json::Value user2 = db->get_user(msgTo);

        if ((user["username"] != username) || (user2["username"] != msgTo)){
            ApiError::setError(response,500,"Internal server error");
        } else {
            if (!db->add_msg(username,user["name"].asString(), msgTo, message)) {
                ApiError::setError(response,500,"Internal server error");
            } else {
                response[STATUS] = SUCCES;
                response[DATA] = OKK;
                //enviar notificacion
                notSender->send_notification(username, user2["firebase_token"].asString(), message);
            }
        }
    } catch (TokenInvalidException &e) {
        ApiError::setError(response,501,"Token invalido");
    } catch (KeyDoesntExistException &e) {
        ApiError::setError(response,500,"Usuario inexistente");
    }
    LOG(INFO) << "USER CHAT POST RESPONSE:\n"
              << "\t\tUser: " << username << "\n"
              << "\t\tResponse: " << response
              << std::endl;
}

void ChatController::deleteConversationRequest(Mongoose::Request &request, Mongoose::JsonResponse &response) {
    std::string token = request.getHeaderKeyValue("Token");
    std::string username;

    try {
        username = sessionManager->get_username(token);

        std::string username2 = request.get("user","");

        LOG(INFO) << "USER CHAT DELETE REQUEST:\n"
                  << "\t\tHeader Token: " << token << "\n"
                  << "\t\tUser: " << username
                  <<"\t\tConversacion con: " << username2 << std::endl;

        Json::Value user = db->get_user(username);
        Json::Value user2 = db->get_user(username2);

        if ((user["username"] != username) || (user2["username"] != username2)) {
            ApiError::setError(response,500,"Internal server error");
        } else {
            if (!db->delete_conv(username, username2)) {
                ApiError::setError(response, 500, "Internal server error");
            } else {
                response[STATUS] = SUCCES;
                response[DATA] = OKK;
            }
        }
    } catch (TokenInvalidException &e) {
        ApiError::setError(response,501,"Token invalido");
    } catch (KeyDoesntExistException &e) {
        ApiError::setError(response,500,"Usuario inexistente");
    }
    LOG(INFO) << "USER CHAT DELETE RESPONSE:\n"
              << "\t\tUser: " << username << "\n"
              << "\t\tResponse: " << response
              << std::endl;
}

void ChatController::deleteMessageRequest(Mongoose::Request &request, Mongoose::JsonResponse &response) {
    std::string token = request.getHeaderKeyValue("Token");
    std::string username;
    std::string data;

    try {
        username = sessionManager->get_username(token);

        data = request.getData();
        Json::Reader reader;
        Json::Value content;
        bool parsingSuccessful = reader.parse(data, content);
        if (!parsingSuccessful) {
            ApiError::setError(response,410,"Wrong JSON");
        }
        std::string username2 = request.get("user","");
        std::string idMessage = content.get("id","").asString();

        LOG(INFO) << "USER MESSAGE DELETE REQUEST:\n"
                  << "\t\tHeader Token: " << token << "\n"
                  << "\t\tUser: " << username
                  <<"\t\tConversacion con: " << username2
                  <<"\t\tMensaje ID: " << idMessage << "-------" << std::endl;

        Json::Value user = db->get_user(username);
        Json::Value user2 = db->get_user(username2);

        if ((user["username"] != username) || (user2["username"] != username2)){
            ApiError::setError(response,500,"Internal server error");
        } else {
            if (!db->delete_msg(username, username2, idMessage)) {
                ApiError::setError(response,500,"Mensaje no encontrado");
            } else {
                response[STATUS] = SUCCES;
                response[DATA] = OKK;
            }
        }
    } catch (TokenInvalidException &e) {
        ApiError::setError(response,501,"Token invalido");
    } catch (KeyDoesntExistException &e) {
        ApiError::setError(response,500,"Usuario inexistente");
    }
    LOG(INFO) << "USER MESSAGE DELETE RESPONSE:\n"
              << "\t\tUser: " << username << "\n"
              << "\t\tResponse: " << response
              << std::endl;
}
